### ISSUE TITLE
Formalize packed format for DAG-CBOR

### ISSUE THREAD

**Issue by burdiyan** [ipld/ipld#337]
CBOR has an upcoming packed format [specification](https://datatracker.ietf.org/doc/draft-ietf-cbor-packed/), the idea of which is to be able to take any CBOR-encoded data, and "compress" it by extracting common map keys into a separate lookup table (stored along with the packed data), and replace map keys with an index into this lookup table.

This is especially helpful when encoding lists of maps with repeated keys, which is highly common. It's technique that's used a lot in many different binary formats, but this spec generalizes it for any CBOR data. Of course this technique is not a replacement for a general-purpose compression algorithm, but it's quite useful for encoding lists of maps. It seems to be a nice middle ground between non-self-describing formats like Protobuf, and very verbose formats self-describing formats like JSON.

Now, one could in theory do it "by hand", and store this lookup table as a plain CBOR list, but the problem is that DAG-CBOR doesn't allow numbers as map keys.

So, I propose to formalize packed representation for DAG-CBOR, probably by referring to the specification above + applying the already existing restrictions of the DAG-CBOR deterministic encoding. There probably would need to be a separate multicodec (e.g. dag-cbor-packed) and a dedicated encoder/decoder in the IPLD libraries.

The motivation for this comes from our experience working with DAG-CBOR: we have lots of lists of maps, and we could very well use some size-reducing properties of packed CBOR. We currently reduce the problem of repeated map keys by using single-letter keys when we have to have big lists of maps, but of course having human-readable names would be much nicer, especially when inspecting the data with the human eye.

**Comment by bumblefudge** [comment]
Makes sense to me.  How "upcoming" is it? It's in draft 13 and the WG chair is driving it-- is it probably reasonable to assume normative v1 will land around Dublin or Bangkok?
